### DISCUSSION TITLE
Creating custom native modules for third-party service

### DISCUSSION THREAD

**Discussion by swamptoday** [expo/expo#24275] [Archived: Questions] [answered]
I am a novice in both React Native and Expo development (and application development in general), and I am currently working on a mobile application project. My application requires the implementation of a "pay for service" feature, and I need to integrate a third-party service provider that offers separate documentation on how to implement Android and iOS SDKs.

I have primarily used Expo for the development of other parts of my application. However, I have encountered a challenge in the integration of this third-party service because it necessitates the use of native modules, which Expo does not directly support in its managed workflow.

I would like to seek guidance on whether creating native modules is the correct approach to solve this problem. Additionally, I am looking for recommended resources and solutions on how to create these native modules and subsequently integrate them into my Expo project. I would greatly appreciate a detailed explanation, but I would be happy with any assistance or pointers in the right direction.

**Comment by tsapeta** [comment, marked as answer]
Expo does support custom native modules in managed workflow (I'm talking about the development builds, not Expo Go). Some resources that might be useful for you:
- More about Expo Modules https://docs.expo.dev/modules/
- Tutorial on how to create a native module using a third-party library https://docs.expo.dev/modules/third-party-library/ (you'll want to use the `--local` flag in `create-expo-module` command)

**Reply by swamptoday** [reply]
Thanks for fast and helpful response!

I have one more question based on this topic and maybe you can share your thoughts on this. I've been mulling over the decision between using custom native modules and making web requests, then displaying the content via a WebView. I think it is one of decisions that can affect the performance and user experience, so I'd love to hear your thoughts.

In your experience, what would you recommend? I'm aware that there might not be a one-size-fits-all answer, but any guidance you can provide would be helpful.

**Reply by tsapeta** [reply]
As you said, it all depends on your use case and how important the user experience is. Personally, I wouldn't use a WebView, mostly due to bad user experience, sometimes also security matters.

**Reply by swamptoday** [reply]
Okay, thank you a lot!

**Reply by ahmedsweng** [reply]
Hello @tsapeta , I hope you are well, can you please help I am stuck with this problem, my problem is when I build my project locally with 
```tsx
eas build -p android --profile development --local 
```
then when I run the app in my phone the native module that I had created with 
```tsx
npx create-expo-module nearby --local
```
is not packed with the build it shows me 
Error: native module Nearby not found
Am I missing something?

**Reply by Thaigun** [reply]
Hey @ahmedNaoumDev , I'm having the exact same issue. I created a module following the tutorial, but it does not work in the built app.

**Comment by yoryi** [comment]
como puedo hacer un modulo nativo con expo que me ayude a subir imagenes en background ??
